src/arraymancer/linear_algebra/special_matrices.nim

```diff
@@ -286,6 +286,174 @@ proc tri*[T](shape_ax1, shape_ax0: int, k: static int = 0, upper: static bool =
 # and which are implemented in helpers/triangular.nim
 export tril, triu
 
+proc circulant*[T](t: Tensor[T], axis = -1, step = 1): Tensor[T] {.noInit.} =
+  ## Construct a circulant matrix from a rank-1 tensor
+  ##
+  ## A circulant matrix is a square matrix in which each column (or row) is
+  ## a cyclic shift of the previous column (or row).
+  ##
+  ## By default this function cirulates over the columns of the output which
+  ## are rotated down by 1 element over the previous column, but this behavior
+  ## can be changed by using the `axis` and `step` arguments.
+  ##
+  ## Inputs:
+  ## - A rank-1 Tensor
+  ## - axis: The axis along which the circulant matrix will be constructed.
+  ##         Defaults to -1 (i.e. the columns, which are the last axis).
+  ## - step: The number of elements by which the input tensor will be shifted
+  ##         each time. Defaults to 1.
+  ## Result:
+  ## - The constructed circulant matrix
+  ##
+  ## Example:
+  ## ```nim
+  ## echo circulant([1, 3, 6])
+  ## Tensor[system.int] of shape "[3, 3]" on backend "Cpu"
+  ## # |1      6     3|
+  ## # |3      1     6|
+  ## # |6      3     1|
+  ## ```
+  result = newTensor[T](t.len, t.len)
+  if axis == 0:
+    for n in 0 ..< t.len:
+      result[n, _] = t.roll(step * n)
+  else:
+    for n in 0 ..< t.len:
+      result[_, n] = t.roll(step * n)
+
+proc toeplitz*[T](c, r: Tensor[T]): Tensor[T] {.noInit.} =
+  ## Construct a Toeplitz matrix
+  ##
+  ## A Toeplitz matrix has constant diagonals, with c as its first column
+  ## and r as its last row (note that `r[0]` is ignored but _should_ be the
+  ## same as `c[^1])`. This is similar to (but different than) a Hankel matrix,
+  ## which has constant anti-diagonals instead.
+  ##
+  ## Inputs:
+  ## - c: The first column of the Toeplitz matrix
+  ## - r: The last row of the Toeplitz matrix (note that `r[0]` is ignored)
+  ## Result:
+  ## - The constructed Toeplitz matrix
+  ##
+  ## Notes:
+  ## - There is a version of this procedure that takes a single argument `c`,
+  ##   in which case `r` is set to `c.conjugate`.
+  ## - Use the `hankel` procedure to generate a Hankel matrix instead.
+  ##
+  ## Example:
+  ## ```nim
+  ## echo toeplitz([1, 3, 6], [9, 10, 11, 12])
```

Comment:
Is that normal that 9 doesn't appear in the resulting example?

Reply:
Yes, actually it is normal that 9 does not appear. The example tries to highlight that r[0] is ignored and that while it should be the same as c[0], that is not enforced. If you think that makes the example confusing I could change it.